### PR TITLE
Add segmented array type

### PR DESCRIPTION
**Pull request by MarcelKoch**
This PR adds a class to store a segmented array. A flat ginkgo array is partitioned into multiple segments by an additional index offset array. The segment `i` starts within the flat buffer at index `offsets[i]`, and ends (exclusively) at index `offsets[i + 1]`. The class only provides access to the flat buffer and the offsets.

Used in: #1543 

PR Stack:

- [ ] #1545 
- [ ] #1543
- [ ] #1579
- [ ] #1544 
- [ ] #1588  
- [ ] #1589 

---
#### Old Description

This PR adds a type `collection::array` which is a `std::vector` like class that stores multiple arrays. These arrays are all subviews of a shared buffer.

I think technically this could be considered as a `batch::array`, but I'm not sure if we want to go that way. As a `batch::array` this would interleave the dependencies between the batched and non-batched stuff, and I'm not sure if that is a good idea.

### CODE REVIEW

**Comment by upsj** [comment]
Do you actually need these individual sub-ranges to be materialized as `array` views? Because otherwise this is just the typical row_ptr - col_idxs structure, and we might be better served with a fully functional `std::span` replacement and a range-of-ranges interface

**Comment by MarcelKoch** [comment]
At least for CPU kernels the arrays are quite handy. But you're right, something like `std::span` would be better suited.

**Comment by upsj** [comment]
I've had it in my mind to provide a nice abstraction over these segmented arrays (including range-for support) for a while already, if you want I can resurrect my prototype code

**Comment by MarcelKoch** [comment]
Sure, if it's not too much work, we could use that. I can also take that over if you want.

**Comment by MarcelKoch** [comment]
@upsj maybe this could be formulated using ranges and accessors. Accessing the segmented array could be thought of as `[i][j]`, where `i` is the array index, and `j` the index within the array. But I don't know enough of ranges and accessor to judge if that would be appropriate.

File: include/ginkgo/core/base/collection.hpp
Comment (upsj) [inline include/ginkgo/core/base/collection.hpp:24]: I think I would call this a _segmented array_, since the corresponding operations are referred to a segmented sort/reduction/... in papers and the cub library. That would also avoid the need for a separate namespace

File: include/ginkgo/core/base/collection.hpp
Comment (upsj) [inline include/ginkgo/core/base/collection.hpp:95]: I think this design might encourage poor performance - it invites using a kernel invocation for each MPI rank, which might scale poorly. Instead I would try to do this with a single kernel for all ranks, and a corresponding device view.

File: include/ginkgo/core/base/collection.hpp
Comment (upsj) [inline include/ginkgo/core/base/collection.hpp:128]: Considering #430, I would prefer if we used int64 or even a generic integer type.

**Review by upsj — COMMENTED** [review]
I think this is a useful construct, I would try to make it more generic in some contexts (index type) and more restricted in others (host-accessibility). I would imagine a device adapter based on what I implemented as `segmented_range` in #1582 might be useful here.

Also we would need some tests and documentation before we can merge this

**Review by pratikvn — COMMENTED** [review]
I agree with Tobias that calling it `segmented_array` would be better. Otherwise, I dont have any other concerns.

File: core/base/segmented_array.cpp
Comment (upsj) [inline core/base/segmented_array.cpp:25]: static_cast to size_type

File: core/base/segmented_array.cpp
Comment (upsj) [inline core/base/segmented_array.cpp:50]: offsets_.size() should never be 0

File: core/base/segmented_array.cpp
Comment (upsj) [inline core/base/segmented_array.cpp:62]: same here

File: include/ginkgo/core/base/segmented_array.hpp
Comment (upsj) [inline include/ginkgo/core/base/segmented_array.hpp:144]: same here, also should that be segment_count?

File: core/base/segmented_array.cpp
Comment (MarcelKoch) [reply core/base/segmented_array.cpp:50]: I was also thinking that. I decided against that, because then I would need to allocate a new array during a move. Do you think that would be acceptable?

**Comment by MarcelKoch** [comment]
@upsj I don't think this will change with the segmented ranges at all. The ranges are only relevant on the backend side. The `index_map` has this type as constructor parameter, and accessors to it, so I would not put it into the detail namespace.

File: include/ginkgo/core/base/segmented_array.hpp
Comment (MarcelKoch) [reply include/ginkgo/core/base/segmented_array.hpp:151]: I'm getting a bit tired of all the rebasing, would it be ok to leave the name as it is?

**Comment by pratikvn** [comment]
Maybe also update the PR name ?

File: include/ginkgo/core/base/segmented_array.hpp
Comment (MarcelKoch) [reply include/ginkgo/core/base/segmented_array.hpp:40]: you're right. I don't remember why I used `std::vector`. I will also use `create_from_*` functions instead.

File: include/ginkgo/core/base/segmented_array.hpp
Comment (MarcelKoch) [reply include/ginkgo/core/base/segmented_array.hpp:5]: These update-every-file changes are always a bit annoying. I would pick up #1466 after the release, perhaps this could be addressed then to keep the disruption minimal.

File: include/ginkgo/core/base/segmented_array.hpp
Comment (MarcelKoch) [reply include/ginkgo/core/base/segmented_array.hpp:151]: changed it, since there were other updates anyway.

**Review by pratikvn — COMMENTED** [review]
Mostly looks good. Just one question regarding the `_const_x` naming for const getters.

File: include/ginkgo/core/base/segmented_array.hpp
Comment (MarcelKoch) [reply include/ginkgo/core/base/segmented_array.hpp:243]: do you mean the `copy_back_deleter`? For that one it makes sense, but the temorary clone implementations are too different.

File: include/ginkgo/core/base/segmented_array.hpp
Comment (MarcelKoch) [reply include/ginkgo/core/base/segmented_array.hpp:121]: How about I keep the two overloads and just add `get_const_*`? For me the overloaded versions are much easier to use in generic programming, so I would prefer to keep them.

File: include/ginkgo/core/base/segmented_array.hpp
Comment (pratikvn) [reply include/ginkgo/core/base/segmented_array.hpp:243]: Yes, that is true. Just for the `copy_back_deleter`.

File: include/ginkgo/core/base/segmented_array.hpp
Comment (pratikvn) [reply include/ginkgo/core/base/segmented_array.hpp:121]: Yes, I am okay with that.

Comment (upsj) [reply include/ginkgo/core/base/segmented_array.hpp:121]: Don't want to block this, but do you need both a mutable and non-mutable version in the same generic code? I would assume in that case it should also be clear whether the object is `const` or not?

File: core/base/segmented_array.cpp
Comment (upsj) [reply core/base/segmented_array.cpp]: Our moves are mostly not `noexcept` anyways, because we can have cross-executor copies when move-assigning. Performance-wise, allocating <= 8 bytes shouldn't really matter, unless we do really weird things.

**Review by upsj — APPROVED** [review]

**Review by pratikvn — APPROVED** [review]

File: include/ginkgo/core/base/segmented_array.hpp
Comment (pratikvn) [inline include/ginkgo/core/base/segmented_array.hpp:159]: You can move this directly within the class, right ?

Comment (MarcelKoch) [reply include/ginkgo/core/base/segmented_array.hpp]: I can actually move them into the .cpp. I did that, but I won't wait for the full CI pipeline to merge it.

**Pull request merged**
